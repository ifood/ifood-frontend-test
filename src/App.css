.App {
  text-align: center;
}

html, body {
  background-color: white;
}

body {
  padding: 10px;
}

