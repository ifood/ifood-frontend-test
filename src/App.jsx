/* eslint-disable react/prefer-stateless-function */
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
import { Provider } from 'react-redux';

import { Component } from 'react';
import Home from './pages/Home';
import Login from './pages/Login';

import store from './store';

import GlobalStyles from './styles/globalStyle';

class App extends Component {
  render() {
    return (
      <Provider store={store}>
        <GlobalStyles />
        <Router>
          <Switch>
            <Route path="/" exact component={Login} />
            <Route path="/home" component={Home} />
          </Switch>
        </Router>
      </Provider>
    );
  }
}

export default App;
