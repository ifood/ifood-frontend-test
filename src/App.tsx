import React from 'react';
import { AppAuthorized } from './AppAuthorized';
import { AppUnauthorized } from './AppUnauthorized';
import { useAuth } from './hooks/useAuth';

export const App: React.FC = () => {
  const { auth } = useAuth();
  return auth ? <AppAuthorized /> : <AppUnauthorized />;
};
