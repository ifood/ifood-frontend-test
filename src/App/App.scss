.app-component {
  display: flex;
  flex-direction: column;
  width: 100%;
  max-height: 100vh;
}
