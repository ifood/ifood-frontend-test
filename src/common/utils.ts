export function authorizeSpotifyUrl(clientId: string | undefined): string {
  return `https://accounts.spotify.com/authorize?client_id=${clientId}&response_type=token&redirect_uri=${process.env.REACT_APP_BASE_API}${process.env.REACT_APP_REDIRECT_AUTHORIZED_URL}`;
}
