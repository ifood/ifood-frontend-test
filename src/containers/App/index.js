import React from 'react'
import { Switch, Route } from 'react-router-dom'
import { ThemeProvider } from 'styled-components'

import HomePage from '../HomePage'
import NotFoundPage from '../NotFoundPage'
import PlaylistsPage from '../PlaylistsPage'

import Header from '../../components/Header'
import Footer from '../../components/Footer'

import { HOME_PATH } from './urls'
import './styles.css'
import { theme } from './theme'

function App() {
  return (
    <ThemeProvider theme={theme}>
      <div className="App">
        <Header />

        <div className="App-container">
          <Switch>
            <Route exact path="/" component={HomePage} />
            <Route path={HOME_PATH} component={PlaylistsPage} />
            <Route component={NotFoundPage} />
          </Switch>
        </div>

        <Footer />
      </div>
    </ThemeProvider>
  )
}

export default App
