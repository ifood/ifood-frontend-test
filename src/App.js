import React, { Component } from 'react';
import './App.css';
import PlaylistCounter from './components/playlist-counter/playlist-counter';
import TotalTracks from './components/total-tracks/total-tracks';
import Filter from './components/filter/filter';
import Playlist from './components/playlist/playlist';
import { URL_ME, URL_FEATURED_PLAYLIST, URL_HEROKU, TOKEN_LOCAL_STORAGE,
  URL_AUTHORIZE, INTERVAL, SPOTIFY_API_ERROR_MESSAGE } from './constants';
import logo from './images/SpotifoodIcon.jpg';
import Axios from 'axios';

class App extends Component {
  localeTemp = '';
  countryTemp = '';
  datetimeTemp = '';
  limitTemp = '';
  offsetTemp = '';

  constructor() {
    super();
    this.state = {
      filterString: '',
      intervalId: null,
      filters: {
        locale: '',
        country: '',
        datetime: '',
        limit: '',
        offset: ''
      }
    }
  }

  // Unmount cycle life
  componentWillUnmount() {
    clearInterval(this.state.intervalId);
  }

  // Method to call Spotify API with filter applied
  callSpotifyPlaylistService = (filtersTemp) => {
    if (!filtersTemp) { // This check is for refresh page
      filtersTemp = this.state.filters;
    }

    let accessToken = localStorage.getItem(TOKEN_LOCAL_STORAGE); // Get the token from local storage
    if (accessToken) {
      // Mount the url filter
      const urlFilterFinal = this.buildFilterUrl(filtersTemp);

      // Get the new Playlist using the filter fields
      this.getPlaylistFromSpotify(urlFilterFinal, accessToken);
     }
  }

   // Method to build the filter url for spotify api
  buildFilterUrl(filtersTemp) {
    let urlFilter = URL_FEATURED_PLAYLIST;
    let hasAtLeastOne = false;

    if (filtersTemp.country) {
      let ctry = "";
      if (filtersTemp.country === "en_US") { // This is a bug in mocky API
        ctry = "US";
      } else {
        ctry = filtersTemp.country;
      }
      urlFilter = urlFilter + "?country=" + ctry;
      hasAtLeastOne = true;
    }

    if (filtersTemp.locale) {
      if (hasAtLeastOne) {
        urlFilter = urlFilter + "&locale=" + filtersTemp.locale;
      } else {
        urlFilter = urlFilter + "?locale=" + filtersTemp.locale;
      }
      hasAtLeastOne = true;
    }

    if (filtersTemp.datetime) {
      let pattern = ":";
      const properlyTimestamp = filtersTemp.datetime.replace(pattern, "%3A");
      if (hasAtLeastOne) {
        urlFilter = urlFilter + "&timestamp=" + properlyTimestamp + "%3A00";
      } else {
        urlFilter = urlFilter + "?timestamp=" + properlyTimestamp + "%3A00";
      }
      hasAtLeastOne = true;
    }

    if (filtersTemp.limit && parseInt(filtersTemp.limit, 10) >= 1 && parseInt(filtersTemp.limit, 10) <= 50) {
      if (hasAtLeastOne) {
        urlFilter = urlFilter + "&limit=" + filtersTemp.limit;
      } else {
        urlFilter = urlFilter + "?limit=" + filtersTemp.limit;
      }
      hasAtLeastOne = true;
    }

    if (filtersTemp.offset && parseInt(filtersTemp.offset, 10) >= 1) {
      if (hasAtLeastOne) {
        urlFilter = urlFilter + "&offset=" + filtersTemp.offset;
      } else {
        urlFilter = urlFilter + "?offset=" + filtersTemp.offset;
      }
    }
    return urlFilter;
  }

  // Method to refresh page after 30 seconds
  refreshPage = () => {
    const intervalId = setInterval(this.callSpotifyPlaylistService, INTERVAL);
    this.setState({
      intervalId,
    });
  }

  // Method called when the user change some information on filter
  filterList = (event) => {
    const field = {
      [event.target.name]: event.target.value,
    };

    if (field.locale) {
      this.localeTemp = field.locale;
    } else if (field.country) {
      this.countryTemp = field.country;
    } else if (field.datetime || field.datetime === "") {
      this.datetimeTemp = field.datetime;
    } else if (field.limit) {
      if (parseInt(field.limit, 10) < 1 || parseInt(field.limit, 10) > 50) {
        this.limitTemp = "";
      } else {
        this.limitTemp = field.limit;
      }
    } else {
      if (parseInt(field.offset, 10) < 1) {
        this.offsetTemp = "";
      } else {
        this.offsetTemp = field.offset;
      }
    }

    this.setState({
      filters: {
        locale: this.localeTemp ? this.localeTemp : '',
        country: this.countryTemp ? this.countryTemp : '',
        datetime: this.datetimeTemp ? this.datetimeTemp : '',
        limit: this.limitTemp ? this.limitTemp : '',
        offset: this.offsetTemp ? this.offsetTemp : ''
      }
    });

    const filtersTemp = {
      locale: this.localeTemp ? this.localeTemp : '',
      country: this.countryTemp ? this.countryTemp : '',
      datetime: this.datetimeTemp ? this.datetimeTemp : '',
      limit: this.limitTemp ? this.limitTemp : '',
      offset: this.offsetTemp ? this.offsetTemp : ''
    };

    this.callSpotifyPlaylistService(filtersTemp);
  }

  // Method to call Spotify Login in case the user is not logged already
  openSpotifyLogin() {
      let url = URL_AUTHORIZE;
          url += '?client_id=' + encodeURIComponent(process.env.REACT_APP_SPOTIFY_CLIENT_ID);
          url += '&response_type=token';
          url += '&redirect_uri=' + encodeURIComponent(URL_HEROKU);
      window.location = url;
  }

  getUserNameInformationFromSpotify(accessToken) {
    Axios.get(URL_ME, {
      headers: { 'Authorization': 'Bearer ' + accessToken }
    }).then(response =>
        this.setState({
        user: {
          name: response.data.display_name
          }
        })
    )
    .catch(error => {
        console.log(SPOTIFY_API_ERROR_MESSAGE, error)
    });
  }

  getPlaylistFromSpotify(url, accessToken) {
    Axios.get(url, {
      headers: { 'Authorization': 'Bearer ' + accessToken }
    }).then(playlistData => {
        if (!playlistData.error) {
        let playlists = playlistData.data.playlists.items;
          this.setState({
            playlists: playlists.map(item => {
              return {
                name: item.name,
                imageUrl: item.images[0].url,
                numberOfTracks: item.tracks.total,
                owner: item.owner.display_name
              }
            })
          })
        } else {
          this.setState({
            playlists: []
          })
        }
      })
      .catch(error => {
        this.setState({
            playlists: []
          })
           console.log(SPOTIFY_API_ERROR_MESSAGE, error)
      });
  }

  // Method called before Render the page
  componentDidMount() {
    const url = window.location.hash;
    if (url) { // Check if the access token is on url
      const access_token = url.match(/\#(?:access_token)\=([\S\s]*?)\&/)[1]; // Get only the access token from url
      localStorage.setItem(TOKEN_LOCAL_STORAGE, access_token); // Save the token on local storage
      window.location.href = '/'; // Clean the URL to not be exposed to user (access token)
    }

    let accessToken = localStorage.getItem(TOKEN_LOCAL_STORAGE); // Check if already have a token into local storage
    if (!accessToken) {
      return;
    }

    // Call this API to get user name information
    this.getUserNameInformationFromSpotify(accessToken);

    // Get all Feature Playlists without filter
    this.getPlaylistFromSpotify(URL_FEATURED_PLAYLIST, accessToken);

    // Start the an interval of 30 seconds to refresh the playlist
    this.refreshPage();
  }

  // Method to render the page
  render() {
    // If there is no user and playlist, return empty array
    let playlistToShow = this.state.user && this.state.playlists ?
      this.state.playlists.filter(playlist => {
        // Here I apply the filter for Playlist name
        let matchesPlaylist = playlist.name.toLowerCase()
        .includes(this.state.filterString.toLocaleLowerCase());
        return matchesPlaylist;
      }) : []

    return (
      <div className="App">
        {
          this.state.user ?
            <div>
              <img src={logo} alt="Spotifood"></img>
              <h2>User logged in: <span style={{color: 'green', fontWeight: 'bold'}}>{this.state.user.name}</span></h2>
              <PlaylistCounter playlists={playlistToShow} />
              <TotalTracks playlists={playlistToShow} />
              <Filter onTextChange={text => this.setState({ filterString: text })}
                onChangeFilters={this.filterList}
                filterValues={this.state.filters} />
              {
                playlistToShow.length > 0 ?
                  playlistToShow.map((playlist, i) => <Playlist key={i} playlist={playlist} index={i} />)
                  : <div style={{ color: 'red', fontWeight: 'bold', marginTop: '50px' }}>No Playlist</div>
              }
            </div> : <button onClick={() => {
                this.openSpotifyLogin();
            }
            }
              style={{ padding: '', fontSize: '50px', marginTop: '20px' }}>Sign in with Spotify</button>
        }
      </div>
    );
  }

}

export default App;