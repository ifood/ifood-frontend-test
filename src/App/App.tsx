import React from 'react';
import { BrowserRouter, Route, Redirect, RouteComponentProps } from 'react-router-dom';

import './App.scss';

import LandingPage from '../pages/LandingPage/LandingPage';
import { SpotifyServiceAuth } from '../services/auth.service';
import Playlists from '../pages/Playlists/Playlists';

const PrivateRoute = ({ Component, path }: { Component: React.FC<RouteComponentProps>, path: string }) => (
  <Route path={path} render={props => (
    SpotifyServiceAuth.hasToken() ? <Component {...props} /> : <Redirect to="/" />
  )} />
);

const App = () => {
  return (
    <div className="app-component">
      <BrowserRouter>
        <Route path="/" exact component={LandingPage} />
        <PrivateRoute path="/playlists" Component={Playlists} />
      </BrowserRouter>
    </div>
  );
}

export default App;
