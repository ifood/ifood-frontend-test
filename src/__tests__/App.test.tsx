import React from 'react';
import { render, screen } from '@testing-library/react';
import { mocked } from 'ts-jest/utils';
import { App } from '../App';
import { AppProvider } from '../contexts/AppContext';
import { useAuth } from '../hooks/useAuth';

jest.mock('../hooks/useAuth');
const useAuthMocked = mocked(useAuth, true);

describe('App', () => {
  beforeEach(() => {
    useAuthMocked.mockClear();
  });

  test('renders the app (unauthorized)', () => {
    useAuthMocked.mockImplementation(() => ({
      auth: null,
      login: jest.fn(),
      logout: jest.fn(),
    }));

    render(<App />, { wrapper: AppProvider });

    expect(screen.getByRole('button', { name: /login/i })).toBeInTheDocument();
    expect(useAuthMocked).toHaveBeenCalled();
  });

  test('renders the app (authorized)', () => {
    useAuthMocked.mockImplementation(() => ({
      auth: { token: '123' },
      login: jest.fn(),
      logout: jest.fn(),
    }));

    render(<App />, { wrapper: AppProvider });

    expect(screen.getByRole('button', { name: /logout/i })).toBeInTheDocument();
    expect(useAuthMocked).toHaveBeenCalled();
  });

  test('matches the snapshot (app unauthorized)', () => {
    useAuthMocked.mockImplementation(() => ({
      auth: null,
      login: jest.fn(),
      logout: jest.fn(),
    }));

    const { baseElement } = render(<App />, { wrapper: AppProvider });
    expect(baseElement).toMatchSnapshot();
  });

  test('matches the snapshot (app authorized)', () => {
    useAuthMocked.mockImplementation(() => ({
      auth: { token: '123' },
      login: jest.fn(),
      logout: jest.fn(),
    }));

    const { baseElement } = render(<App />, { wrapper: AppProvider });
    expect(baseElement).toMatchSnapshot();
  });
});
