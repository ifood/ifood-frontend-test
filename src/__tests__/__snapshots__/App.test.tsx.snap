// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches the snapshot (app authorized) 1`] = `
<body>
  <div>
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-dense MuiToolbar-gutters"
      >
        <h2
          class="MuiTypography-root Header-title-16 MuiTypography-h2"
        >
          Spotifood
        </h2>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Logout
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </header>
    <div
      class="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthMd"
    >
      <div
        class="MuiPaper-root Filter-container-18 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="Filter-search-19"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart MuiInputBase-marginDense MuiOutlinedInput-marginDense"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-marginDense"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorAction"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
              </div>
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
                placeholder="Buscar playlists..."
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-23 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-24"
                  style="width: 0.01px;"
                >
                  <span>
                    â€‹
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
                  />
                </svg>
              </span>
              Filtros
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiCollapse-container MuiCollapse-hidden"
          style="min-height: 0px;"
        >
          <div
            class="MuiCollapse-wrapper"
          >
            <div
              class="MuiCollapse-wrapperInner"
            >
              <div
                class="Filter-indicator-22"
              >
                <div
                  class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
                  role="progressbar"
                  style="width: 40px; height: 40px;"
                >
                  <svg
                    class="MuiCircularProgress-svg"
                    viewBox="22 22 44 44"
                  >
                    <circle
                      class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                      cx="44"
                      cy="44"
                      fill="none"
                      r="20.2"
                      stroke-width="3.6"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="FeaturedPlaylists-indicator-17"
      >
        <div
          class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`App matches the snapshot (app unauthorized) 1`] = `
<body>
  <div>
    <div
      class="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthMd"
    >
      <div
        class="Login-container-14"
      >
        <h2
          class="MuiTypography-root Login-title-15 MuiTypography-h2"
        >
          Spotifood
        </h2>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-disableElevation"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Login
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
